### PR TITLE
[FLINK-20783][table-planner-blink] Separate implementation of ExecNode and PhysicalNode in batch join

### PR DESCRIPTION
**Pull request by wenlong88**
## What is the purpose of the change
Separate implementation of HashJoin/NestedLoopJoin/SortMergeJoin in batch

## Brief change log
1. Introduce BatchPhysicalHashJoin, and make BatchExecHashJoin only extended from ExecNode
2. Introduce BatchPhysicalNestedLoopJoin, and make BatchExecNestedLoopJoin only extended from ExecNode
3. Introduce BatchPhysicalSortMergeJoin, and make BatchExecSortMergeJoin only extended from ExecNode

## Verifying this change
This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            long rightRowCount,
+            boolean leftIsBuild,
+            boolean tryDistinctBuildRow,
+            List<ExecEdge> inputEdges,
+            LogicalType outputType,
+            String description) {
+        super(inputEdges, outputType, description);
+        this.joinType = joinType;
+
+        Preconditions.checkArgument(leftKeys.length == rightKeys.length);
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java:86]: `leftKeys` should be non-empty.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            bRowSize = rightAvgRowSize;
+            bRowCount = rightRowCount;
+            bKeys = rightKeys;
+
+            probe = lInput;
+            pProj = lProj;
+            pType = lType;
+            pRowCount = leftRowCount;
+            pKeys = leftKeys;
+        }
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java:175]: give build/probe fields directly in the constructor ?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchPhysicalHashJoin.scala [hunk 3]
```diff
@@ -174,7 +163,37 @@ class BatchExecHashJoin(
 
   //~ ExecNode methods -----------------------------------------------------------
 
-  override def getInputEdges: util.List[ExecEdge] = {
+  override def translateToExecNode(): ExecNode[_] = {
+
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchPhysicalHashJoin.scala:167]: remove empty line

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchPhysicalHashJoin.scala [hunk 4]
```diff
@@ -174,7 +163,37 @@ class BatchExecHashJoin(
 
   //~ ExecNode methods -----------------------------------------------------------
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchPhysicalHashJoin.scala:164]: remove this line

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+import java.util.stream.IntStream;
+
+/** Batch {@link ExecNode} for Hash Join. */
+public class BatchExecHashJoin extends ExecNodeBase<RowData> implements BatchExecNode<RowData> {
+
+    private final FlinkJoinType joinType;
+    private final int[] leftKeys;
+    private final int[] rightKeys;
+    private final boolean[] filterNulls;
+    private final RexNode nonEquiConditions;
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java:60]: add `@Nullable` annotation

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    private final long leftRowCount;
+    private final long rightRowCount;
+    private final boolean tryDistinctBuildRow;
+
+    public BatchExecHashJoin(
+            FlinkJoinType joinType,
+            int[] leftKeys,
+            int[] rightKeys,
+            boolean[] filterNulls,
+            RexNode nonEquiConditions,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java:73]: add `@Nullable` annotation

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchPhysicalHashJoin.scala [hunk 7]
```diff
@@ -174,7 +163,37 @@ class BatchExecHashJoin(
...
+        rightRowCount,
+        leftIsBuild,
+        tryDistinctBuildRow,
+        getInputEdges,
+        FlinkTypeFactory.toLogicalRowType(getRowType),
+        getRelDetailedDescription
+    )
+  }
+
+  def getInputEdges: util.List[ExecEdge] = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchPhysicalHashJoin.scala:196]: add `private`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecNestedLoopJoin.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    private final boolean leftIsBuild;
+    private final boolean singleRowJoin;
+
+    public BatchExecNestedLoopJoin(
+            FlinkJoinType joinType,
+            RexNode condition,
+            boolean leftIsBuild,
+            boolean singleRowJoin,
+            List<ExecEdge> inputEdges,
+            LogicalType outputType,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecNestedLoopJoin.java:59]: RowType

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+/** Batch {@link ExecNode} for Sort Merge Join. */
+public class BatchExecSortMergeJoin extends ExecNodeBase<RowData>
+        implements BatchExecNode<RowData> {
+
+    private final FlinkJoinType joinType;
+    private final int[] leftKeys;
+    private final int[] rightKeys;
+    private final boolean[] filterNulls;
+    private final RexNode nonEquiConditions;
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:62]: add @Nullable annotation

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+    public BatchExecSortMergeJoin(
+            FlinkJoinType joinType,
+            int[] leftKeys,
+            int[] rightKeys,
+            boolean[] filterNulls,
+            RexNode nonEquiConditions,
+            boolean leftIsSmaller,
+            List<ExecEdge> inputEdges,
+            LogicalType outputType,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:73]: RowType

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            String description) {
+        super(inputEdges, outputType, description);
+        this.joinType = joinType;
+
+        Preconditions.checkArgument(leftKeys.length == rightKeys.length);
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:78]: `leftKeys` should not be empty

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    private final boolean[] filterNulls;
+    private final RexNode nonEquiConditions;
+    private final boolean leftIsSmaller;
+
+    public BatchExecSortMergeJoin(
+            FlinkJoinType joinType,
+            int[] leftKeys,
+            int[] rightKeys,
+            boolean[] filterNulls,
+            RexNode nonEquiConditions,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:70]: add @Nullable annotation

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+        Transformation<RowData> lInput =
+                (Transformation<RowData>) getInputNodes().get(0).translateToPlan(planner);
+        Transformation<RowData> rInput =
+                (Transformation<RowData>) getInputNodes().get(1).translateToPlan(planner);
+
+        // get type
+        RowType lType = (RowType) getInputNodes().get(0).getOutputType();
+        RowType rType = (RowType) getInputNodes().get(1).getOutputType();
+
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:100]: add left/right input node local variable ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+                                                ExecutionConfigOptions
+                                                        .TABLE_EXEC_RESOURCE_EXTERNAL_BUFFER_MEMORY))
+                        .getBytes();
+        long sortMemory =
+                MemorySize.parse(
+                                config.getConfiguration()
+                                        .getString(
+                                                ExecutionConfigOptions
+                                                        .TABLE_EXEC_RESOURCE_SORT_MEMORY))
+                        .getBytes();
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:121]: extract an utility method ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+        }
+        return ret;
+    }
+
+    private SortCodeGenerator newSortGen(TableConfig config, int[] originalKeys, RowType type) {
+        boolean[] originalOrders = new boolean[originalKeys.length];
+        Arrays.fill(originalOrders, true);
+        boolean[] nullsIsLast = SortUtil.getNullDefaultOrders(originalOrders);
+        Tuple3<int[], boolean[], boolean[]> ret =
+                SortUtil.deduplicateSortKeys(originalKeys, originalOrders, nullsIsLast);
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:181]: add some comments to explain what's each element in Tuple3?

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @wenlong88 , I left some comments

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java (hunk 2 above)
Comment (wenlong88) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java:175]: the flag leftIsBuild cannot be reduced, so there is no need to specified build/probe side in the constructor, it is better to left them inside the ExecNode.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java (hunk 15 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java:178]: return `SortSpec` directly ? even we can use `SortSpec` in `SortCodeGenerator` ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashJoin.java:90]: I think this validation is necessary.

**Comment by godfreyhe** [comment]
Thanks for the update. LGTM overall, one minor comment: we could also use SortSpec in `ComparatorCodeGenerator`, and please fix the compile error

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java [hunk 16]
```diff
@@ -121,16 +121,24 @@ public StreamExecRank(
...
         int[] sortKeyPositions = IntStream.range(0, sortFields.length).toArray();
+        SortSpec.SortSpecBuilder builder = SortSpec.builder();
+        IntStream.range(0, sortFields.length)
+                .forEach(
+                        idx ->
+                                builder.addField(
+                                        idx,
+                                        sortSpec.getFieldSpec(idx).getIsAscendingOrder(),
+                                        sortSpec.getFieldSpec(idx).getNullIsLast()));
+        SortSpec sortSpecInSortKey = builder.build();
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java:134]: add a `createSubSortSpec(int startIndex, int endIndex)` method in `SortSpec` and replace this code snippet with the new method ?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/sort/ComparatorCodeGenerator.scala [hunk 17]
```diff
@@ -36,25 +37,27 @@ object ComparatorCodeGenerator {
...
     val baseClass = classOf[RecordComparator]
 
     val ctx = new CodeGeneratorContext(conf)
     val compareCode = GenerateUtils.generateRowCompare(
-      ctx, keys, keyTypes, orders, nullsIsLast, "o1", "o2")
+        ctx,
+        sortSpec.getFieldIndices,
+        sortSpec.getFieldTypes(input),
+        sortSpec.getAscendingOrders,
+        sortSpec.getNullsIsLast,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/sort/ComparatorCodeGenerator.scala:58]: replace the sort related parameters with `SortSpec` for `GenerateUtils.generateRowCompare` method ?

**Review by godfreyhe — COMMENTED** [review]
Thanks for the quick update, I left two minor comments about the new commit.

**Comment by godfreyhe** [comment]
Thanks for the contribution, I will merge this pr

**Pull request closed without merging**